logprep/framework/pipeline_manager.py

```diff
@@ -53,6 +57,93 @@ def put(self, obj, block=True, timeout=None, batch_size=1):
         super().put(obj, block=block, timeout=timeout)
 
 
+@define()
+class ComponentQueueListener:
+    """This forks a process and handles all items from the given queue into
+    the specified callable. It uses a sentinel object to stop the process."""
+
+    queue: multiprocessing.queues.Queue = field(
+        validator=validators.instance_of(multiprocessing.queues.Queue)
+    )
+    """The queue to listen to."""
+
+    target: str = field(validator=validators.instance_of(str))
+    """The method name to call with the items from the queue."""
+
+    config: dict = field(validator=validators.instance_of(dict))
+    """The configuration for the listener component."""
+
+    sentinel: Any = field(default=None)
+    """The sentinel object to stop the process. This has to implement identity comparison."""
+
+    _instance: multiprocessing.Process | threading.Thread = field(init=False)
+    """The process that is forked to listen to the queue."""
+
+    _implementation: str = field(
+        default="threading", validator=validators.in_(["threading", "multiprocessing"])
+    )
+    """The implementation to use for the listener. Options are threading or multiprocessing.
+    Default is threading."""
+
+    def __attrs_post_init__(self):
+        if self._implementation == "threading":
+            self._instance = threading.Thread(target=self._listen, daemon=True)
+        elif self._implementation == "multiprocessing":
+            self._instance = multiprocessing.Process(target=self._listen, daemon=True)
+
+    def start(self):
+        """Start the listener."""
+        logger.debug("Starting listener with target: %s", self.target)
+        self._instance.start()
+
+    def get_component_instance(self):
+        component = Factory.create(self.config)
+        try:
+            component.setup()
+            self.queue.put(1)
+        except SystemExit as error:
+            logger.error("Error output not reachable. Exiting...")
+            self.queue.put(self.sentinel)
+            raise error from error
+        # wait for setup method in pipeline manager to receive the message
+        while not self.queue.empty():
+            logger.debug("Waiting for receiver to be ready")
+        return component
+
+    def _listen(self):
+        component = self.get_component_instance()
+        target = getattr(component, self.target)
+        while 1:
+            item = self.queue.get()
+            logger.debug("Got item from queue: %s", item)
+            if item is self.sentinel:
+                logger.debug("Got sentinel. Stopping listener.")
+                break
+            try:
+                target(item)
+            except Exception as error:  # pylint: disable=broad-except
+                logger.error("Error processing item: %s due to %s", item, error)
+        self._drain_queue(target)
+        component.shut_down()
+
+    def _drain_queue(self, target):
+        while not self.queue.empty():
```

Comment:
closing the queue does indeed prevent also taking from the queue so that is not an option. A possibility would be to implement a custom queue, that inherits from the original one, and has it's own closing mechanism that allows reading from the queue, but not adding to it. I decided not to go for this solution here, as the current implementation should suffice for now.